gadget/device_linux.go

```diff
@@ -202,17 +204,47 @@ func findParentDeviceWithWritableFallback() (string, error) {
 	if err != nil {
 		return "", err
 	}
-	return ParentDiskFromPartition(partitionWritable)
+	return ParentDiskFromMountSource(partitionWritable)
 }
 
-// ParentDiskFromPartition will find the parent disk device for the
-// given partition. E.g. /dev/nvmen0n1p5 -> /dev/nvme0n1
+// ParentDiskFromMountSource will find the parent disk device for the given
+// partition. E.g. /dev/nvmen0n1p5 -> /dev/nvme0n1.
 //
-// Note that this does not work for anything using device mapper (like
-// LUKS/LVM) yet.
-func ParentDiskFromPartition(partition string) (string, error) {
+// When the mount source is a symlink, it is resolved to the actual device that
+// is mounted. Should the device be one created by device mapper, it is followed
+// up to the actual underlying block device. As an example, this is how devices
+// are followed with a /writable mounted from an encrypted volume:
+//
+// /dev/mapper/ubuntu-data-<uuid> (a symlink)
+//   ⤷ /dev/dm-0 (set up by device mapper)
+//       ⤷ /dev/hda4 (actual partition with the content)
+//          ⤷ /dev/hda (returned by this function)
+//
+func ParentDiskFromMountSource(mountSource string) (string, error) {
```

Comment:
this function is very similar in scope to the existing `disks.DiskFromMountpoint` but seems far simpler, I wonder if we could either replace that function with this one or replace this function with that one, I don't see exactly where this one is necessarily better or worse compared to the one that I wrote, but it would be nice to figure out which one is better and consolidate the two

Reply:
Fair point, something I can take a look at today.